File: supabase/functions/chatbot-whatsapp-webhook/chatbot/billing_management.ts
import { ConversationState, updateConversationState } from './conversation_state.ts'
import { 
  getCustomerPaymentMethods,
  deletePaymentMethod,
  setDefaultPaymentMethod,
  getCustomerProfile
} from './database_helpers.ts'
import { 
  formatBillingManagement,
  formatPaymentMethodActions,
  formatPaymentMethodDeleted,
  formatPaymentMethodSetDefault,
  formatAddPaymentMethodEmailRequired,
  formatAddPaymentMethod,
  formatPaymentMethodAdded,
  formatMainMenu,
  formatProfileView,
  formatError 
} from './message_formatter.ts'

export async function handleBillingManagement(
  chatId: string,
  customerId: string,
  messageBody: string,
  state: ConversationState
): Promise<string> {
  switch (state.step) {
    case 'billing_list':
      return await handleBillingList(chatId, customerId, messageBody, state)
      
    case 'billing_actions':
      return await handleBillingActions(chatId, customerId, messageBody, state)
      
    case 'billing_add_email_check':
      return await handleBillingAddEmailCheck(chatId, customerId, messageBody, state)
      
    case 'billing_add':
      return await handleBillingAdd(chatId, customerId, messageBody, state)
      
    case 'billing_added':
      return await handleBillingAdded(chatId, customerId, messageBody, state)
      
    default:
      return formatError('Invalid billing management step')
  }
}

async function handleBillingList(chatId: string, customerId: string, messageBody: string, state: ConversationState): Promise<string> {
  const selectedNumber = parseInt(messageBody.trim())
  
  // Get payment methods
  const paymentMethods = await getCustomerPaymentMethods(customerId)
  
  if (!isNaN(selectedNumber)) {
    // Check if user selected a payment method
    if (selectedNumber >= 1 && selectedNumber <= paymentMethods.length) {
      const selectedMethod = paymentMethods[selectedNumber - 1]
      
      await updateConversationState(chatId, {
        ...state,
        step: 'billing_actions',
        selected_payment_method_id: selectedMethod.id,
        previous_step: 'billing_list'
      })
      
      return formatPaymentMethodActions(selectedMethod)
    }
    
    // Handle options (Add Payment Method, Back to Main Menu)
    const addOptionNumber = paymentMethods.length + 1
    const backOptionNumber = paymentMethods.length + 2
    
    if (selectedNumber === addOptionNumber) {
      // Add Payment Method - check for email first
      const customer = await getCustomerProfile(customerId)
      if (!customer?.email) {
        await updateConversationState(chatId, {
          ...state,
          step: 'billing_add_email_check',
          previous_step: 'billing_list'
        })
        return formatAddPaymentMethodEmailRequired()
      } else {
        await updateConversationState(chatId, {
          ...state,
          step: 'billing_add',
          previous_step: 'billing_list'
        })
        return formatAddPaymentMethod()
      }
    } else if (selectedNumber === backOptionNumber) {
      // Back to Main Menu
      await updateConversationState(chatId, {
        ...state,
        step: 'main_menu',
        previous_step: 'billing_list'
      })
      return formatMainMenu()
    } else {
      // Invalid selection
      const maxOption = paymentMethods.length === 0 ? 2 : paymentMethods.length + 2
      const errorMessage = formatError(`Please select a valid option (1-${maxOption})`)
      const billingView = formatBillingManagement(paymentMethods)
      return errorMessage + '\n\n' + billingView
    }
  }
  
  // Default: show billing list
  return formatBillingManagement(paymentMethods)
}

async function handleBillingActions(chatId: string, customerId: string, messageBody: string, state: ConversationState): Promise<string> {
  const selectedNumber = parseInt(messageBody.trim())
  
  if (!state.selected_payment_method_id) {
    return formatError('Payment method not found. Please try again.')
  }
  
  // Get the payment method details
  const paymentMethods = await getCustomerPaymentMethods(customerId)
  const selectedMethod = paymentMethods.find(method => method.id === state.selected_payment_method_id)
  
  if (!selectedMethod) {
    return formatError('Payment method not found. Please try again.')
  }
  
  const isDefault = selectedMethod.is_default
  
  if (!isNaN(selectedNumber)) {
    if (isDefault) {
      // Default card actions: Remove, Back, Main Menu
      switch (selectedNumber) {
        case 1: // Remove Card
          const deleteResult = await deletePaymentMethod(customerId, state.selected_payment_method_id)
          if (deleteResult.success) {
            const updatedPaymentMethods = await getCustomerPaymentMethods(customerId)
            await updateConversationState(chatId, {
              ...state,
              step: 'billing_list',
              selected_payment_method_id: undefined,
              previous_step: 'billing_actions'
            })
            
            const successMessage = formatPaymentMethodDeleted()
            const billingView = formatBillingManagement(updatedPaymentMethods)
            return successMessage + '\n\n' + billingView
          } else {
            return formatError(deleteResult.error || 'Failed to remove payment method')
          }
          
        case 2: // Back to Payment Methods
          const paymentMethodsBack = await getCustomerPaymentMethods(customerId)
          await updateConversationState(chatId, {
            ...state,
            step: 'billing_list',
            selected_payment_method_id: undefined,
            previous_step: 'billing_actions'
          })
          return formatBillingManagement(paymentMethodsBack)
          
        case 3: // Main Menu
          await updateConversationState(chatId, {
            ...state,
            step: 'main_menu',
            selected_payment_method_id: undefined,
            previous_step: 'billing_actions'
          })
          return formatMainMenu()
          
        default:
          const errorMessage = formatError('Please select a valid option (1-3)')
          const actionsView = formatPaymentMethodActions(selectedMethod)
          return errorMessage + '\n\n' + actionsView
      }
    } else {
      // Non-default card actions: Set Default, Remove, Back, Main Menu
      switch (selectedNumber) {
        case 1: // Set as Default
          const setDefaultResult = await setDefaultPaymentMethod(customerId, state.selected_payment_method_id)
          if (setDefaultResult.success) {
            const updatedPaymentMethods = await getCustomerPaymentMethods(customerId)
            await updateConversationState(chatId, {
              ...state,
              step: 'billing_list',
              selected_payment_method_id: undefined,
              previous_step: 'billing_actions'
            })
            
            const successMessage = formatPaymentMethodSetDefault()
            const billingView = formatBillingManagement(updatedPaymentMethods)
            return successMessage + '\n\n' + billingView
          } else {
            return formatError(setDefaultResult.error || 'Failed to set default payment method')
          }
          
        case 2: // Remove Card
          const deleteResult = await deletePaymentMethod(customerId, state.selected_payment_method_id)
          if (deleteResult.success) {
            const updatedPaymentMethods = await getCustomerPaymentMethods(customerId)
            await updateConversationState(chatId, {
              ...state,
              step: 'billing_list',
              selected_payment_method_id: undefined,
              previous_step: 'billing_actions'
            })
            
            const successMessage = formatPaymentMethodDeleted()
            const billingView = formatBillingManagement(updatedPaymentMethods)
            return successMessage + '\n\n' + billingView
          } else {
            return formatError(deleteResult.error || 'Failed to remove payment method')
          }
          
        case 3: // Back to Payment Methods
          const paymentMethodsBack = await getCustomerPaymentMethods(customerId)
          await updateConversationState(chatId, {
            ...state,
            step: 'billing_list',
            selected_payment_method_id: undefined,
            previous_step: 'billing_actions'
          })
          return formatBillingManagement(paymentMethodsBack)
          
        case 4: // Main Menu
          await updateConversationState(chatId, {
            ...state,
            step: 'main_menu',
            selected_payment_method_id: undefined,
            previous_step: 'billing_actions'
          })
          return formatMainMenu()
          
        default:
          const errorMessage = formatError('Please select a valid option (1-4)')
          const actionsView = formatPaymentMethodActions(selectedMethod)
          return errorMessage + '\n\n' + actionsView
      }
    }
  }
  
  // If not a number, show error with menu
  const errorMessage = formatError('Please select a valid option')
  const actionsView = formatPaymentMethodActions(selectedMethod)
  return errorMessage + '\n\n' + actionsView
}

async function handleBillingAddEmailCheck(chatId: string, customerId: string, messageBody: string, state: ConversationState): Promise<string> {
  const selectedNumber = parseInt(messageBody.trim())
  
  if (!isNaN(selectedNumber)) {
    switch (selectedNumber) {
      case 1: // Go to Profile
        const customer = await getCustomerProfile(customerId)
        if (!customer) {
          return formatError('Unable to load profile. Please try again.')
        }
        
        await updateConversationState(chatId, {
          ...state,
          step: 'profile_view',
          previous_step: 'billing_add_email_check'
        })
        
        return formatProfileView(customer)
        
      case 2: // Back to Payment Methods
        const paymentMethods = await getCustomerPaymentMethods(customerId)
        await updateConversationState(chatId, {
          ...state,
          step: 'billing_list',
          previous_step: 'billing_add_email_check'
        })
        return formatBillingManagement(paymentMethods)
        
      case 3: // Back to Main Menu
        await updateConversationState(chatId, {
          ...state,
          step: 'main_menu',
          previous_step: 'billing_add_email_check'
        })
        return formatMainMenu()
        
      default:
        const errorMessage = formatError('Please select a valid option (1-3)')
        const emailRequiredView = formatAddPaymentMethodEmailRequired()
        return errorMessage + '\n\n' + emailRequiredView
    }
  }
  
  // If not a number, show error with menu
  const errorMessage = formatError('Please select a valid option')
  const emailRequiredView = formatAddPaymentMethodEmailRequired()
  return errorMessage + '\n\n' + emailRequiredView
}

async function handleBillingAdd(chatId: string, customerId: string, messageBody: string, state: ConversationState): Promise<string> {
  const selectedNumber = parseInt(messageBody.trim())
  
  if (!isNaN(selectedNumber)) {
    switch (selectedNumber) {
      case 1: // Get Payment Link
        // TODO: Implement PayStack payment link generation
        // For now, simulate successful addition
        await updateConversationState(chatId, {
          ...state,
          step: 'billing_added',
          previous_step: 'billing_add'
        })
        return formatPaymentMethodAdded()
        
      case 2: // Back to Payment Methods
        const paymentMethods = await getCustomerPaymentMethods(customerId)
        await updateConversationState(chatId, {
          ...state,
          step: 'billing_list',
          previous_step: 'billing_add'
        })
        return formatBillingManagement(paymentMethods)
        
      case 3: // Back to Main Menu
        await updateConversationState(chatId, {
          ...state,
          step: 'main_menu',
          previous_step: 'billing_add'
        })
        return formatMainMenu()
        
      default:
        const errorMessage = formatError('Please select a valid option (1-3)')
        const addPaymentView = formatAddPaymentMethod()
        return errorMessage + '\n\n' + addPaymentView
    }
  }
  
  // If not a number, show error with menu
  const errorMessage = formatError('Please select a valid option')
  const addPaymentView = formatAddPaymentMethod()
  return errorMessage + '\n\n' + addPaymentView
}

async function handleBillingAdded(chatId: string, customerId: string, messageBody: string, state: ConversationState): Promise<string> {
  // After payment method is added, show the updated billing list
  const paymentMethods = await getCustomerPaymentMethods(customerId)
  await updateConversationState(chatId, {
    ...state,
    step: 'billing_list',
    previous_step: 'billing_added'
  })
  return formatBillingManagement(paymentMethods)
} 

File: backend/internal/chatbot/billing_management.go
package chatbot

import (
	"context"
	"fmt"
)

func (s *Service) handleBillingManagement(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	switch state.Step {
	case "billing_list":
		return s.handleBillingList(ctx, chatID, customerID, messageBody, state)
	case "billing_actions":
		return s.handleBillingActions(ctx, chatID, customerID, messageBody, state)
	case "billing_add_email_check":
		return s.handleBillingAddEmailCheck(ctx, chatID, customerID, messageBody, state)
	case "billing_add":
		return s.handleBillingAdd(ctx, chatID, customerID, messageBody, state)
	case "billing_added":
		return s.handleBillingAdded(ctx, chatID, customerID, messageBody, state)
	default:
		return formatError("Invalid billing management step")
	}
}

func (s *Service) handleBillingList(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	selectedNumber, hasNum := parseIntTrim(messageBody)
	paymentMethods := s.getCustomerPaymentMethods(ctx, customerID)

	if hasNum {
		if selectedNumber >= 1 && selectedNumber <= len(paymentMethods) {
			selected := paymentMethods[selectedNumber-1]
			newState := state
			newState.Step = "billing_actions"
			newState.SelectedPaymentMethodID = selected.ID
			newState.PreviousStep = "billing_list"
			s.updateConversationState(ctx, chatID, newState)
			return formatPaymentMethodActions(&selected)
		}
		addOption := len(paymentMethods) + 1
		backOption := len(paymentMethods) + 2

		if selectedNumber == addOption {
			customer := s.getCustomerProfile(ctx, customerID)
			if customer == nil || customer.Email == nil || *customer.Email == "" {
				newState := state
				newState.Step = "billing_add_email_check"
				newState.PreviousStep = "billing_list"
				s.updateConversationState(ctx, chatID, newState)
				return formatAddPaymentMethodEmailRequired()
			}
			newState := state
			newState.Step = "billing_add"
			newState.PreviousStep = "billing_list"
			s.updateConversationState(ctx, chatID, newState)
			return formatAddPaymentMethod()
		} else if selectedNumber == backOption {
			newState := state
			newState.Step = "main_menu"
			newState.PreviousStep = "billing_list"
			s.updateConversationState(ctx, chatID, newState)
			return formatMainMenu("", 0, 0, "")
		}
		maxOption := 2
		if len(paymentMethods) > 0 {
			maxOption = len(paymentMethods) + 2
		}
		return formatError(fmt.Sprintf("Please select a valid option (1-%d)", maxOption)) + "\n\n" + formatBillingManagement(paymentMethods)
	}
	return formatBillingManagement(paymentMethods)
}

func (s *Service) handleBillingActions(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	selectedNumber, hasNum := parseIntTrim(messageBody)
	if state.SelectedPaymentMethodID == "" {
		return formatError("Payment method not found. Please try again.")
	}

	paymentMethods := s.getCustomerPaymentMethods(ctx, customerID)
	var selectedMethod *PaymentMethod
	for i := range paymentMethods {
		if paymentMethods[i].ID == state.SelectedPaymentMethodID {
			selectedMethod = &paymentMethods[i]
			break
		}
	}
	if selectedMethod == nil {
		return formatError("Payment method not found. Please try again.")
	}
	isDefault := selectedMethod.IsDefault

	if hasNum {
		if isDefault {
			switch selectedNumber {
			case 1:
				ok, errMsg := s.deletePaymentMethod(ctx, customerID, state.SelectedPaymentMethodID)
				if ok {
					updated := s.getCustomerPaymentMethods(ctx, customerID)
					newState := state
					newState.Step = "billing_list"
					newState.SelectedPaymentMethodID = ""
					newState.PreviousStep = "billing_actions"
					s.updateConversationState(ctx, chatID, newState)
					return formatPaymentMethodDeleted() + "\n\n" + formatBillingManagement(updated)
				}
				if errMsg == "" {
					errMsg = "Failed to remove payment method"
				}
				return formatError(errMsg)
			case 2:
				updated := s.getCustomerPaymentMethods(ctx, customerID)
				newState := state
				newState.Step = "billing_list"
				newState.SelectedPaymentMethodID = ""
				newState.PreviousStep = "billing_actions"
				s.updateConversationState(ctx, chatID, newState)
				return formatBillingManagement(updated)
			case 3:
				newState := state
				newState.Step = "main_menu"
				newState.SelectedPaymentMethodID = ""
				newState.PreviousStep = "billing_actions"
				s.updateConversationState(ctx, chatID, newState)
				return formatMainMenu("", 0, 0, "")
			default:
				return formatError("Please select a valid option (1-3)") + "\n\n" + formatPaymentMethodActions(selectedMethod)
			}
		}
		switch selectedNumber {
		case 1:
			ok, errMsg := s.setDefaultPaymentMethod(ctx, customerID, state.SelectedPaymentMethodID)
			if ok {
				updated := s.getCustomerPaymentMethods(ctx, customerID)
				newState := state
				newState.Step = "billing_list"
				newState.SelectedPaymentMethodID = ""
				newState.PreviousStep = "billing_actions"
				s.updateConversationState(ctx, chatID, newState)
				return formatPaymentMethodSetDefault() + "\n\n" + formatBillingManagement(updated)
			}
			if errMsg == "" {
				errMsg = "Failed to set default payment method"
			}
			return formatError(errMsg)
		case 2:
			ok, errMsg := s.deletePaymentMethod(ctx, customerID, state.SelectedPaymentMethodID)
			if ok {
				updated := s.getCustomerPaymentMethods(ctx, customerID)
				newState := state
				newState.Step = "billing_list"
				newState.SelectedPaymentMethodID = ""
				newState.PreviousStep = "billing_actions"
				s.updateConversationState(ctx, chatID, newState)
				return formatPaymentMethodDeleted() + "\n\n" + formatBillingManagement(updated)
			}
			if errMsg == "" {
				errMsg = "Failed to remove payment method"
			}
			return formatError(errMsg)
		case 3:
			updated := s.getCustomerPaymentMethods(ctx, customerID)
			newState := state
			newState.Step = "billing_list"
			newState.SelectedPaymentMethodID = ""
			newState.PreviousStep = "billing_actions"
			s.updateConversationState(ctx, chatID, newState)
			return formatBillingManagement(updated)
		case 4:
			newState := state
			newState.Step = "main_menu"
			newState.SelectedPaymentMethodID = ""
			newState.PreviousStep = "billing_actions"
			s.updateConversationState(ctx, chatID, newState)
			return formatMainMenu("", 0, 0, "")
		default:
			return formatError("Please select a valid option (1-4)") + "\n\n" + formatPaymentMethodActions(selectedMethod)
		}
	}
	return formatError("Please select a valid option") + "\n\n" + formatPaymentMethodActions(selectedMethod)
}

func (s *Service) handleBillingAddEmailCheck(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	selectedNumber, hasNum := parseIntTrim(messageBody)
	if hasNum {
		switch selectedNumber {
		case 1:
			customer := s.getCustomerProfile(ctx, customerID)
			if customer == nil {
				return formatError("Unable to load profile. Please try again.")
			}
			newState := state
			newState.Step = "profile_view"
			newState.PreviousStep = "billing_add_email_check"
			s.updateConversationState(ctx, chatID, newState)
			return formatProfileView(customer)
		case 2:
			paymentMethods := s.getCustomerPaymentMethods(ctx, customerID)
			newState := state
			newState.Step = "billing_list"
			newState.PreviousStep = "billing_add_email_check"
			s.updateConversationState(ctx, chatID, newState)
			return formatBillingManagement(paymentMethods)
		case 3:
			newState := state
			newState.Step = "main_menu"
			newState.PreviousStep = "billing_add_email_check"
			s.updateConversationState(ctx, chatID, newState)
			return formatMainMenu("", 0, 0, "")
		default:
			return formatError("Please select a valid option (1-3)") + "\n\n" + formatAddPaymentMethodEmailRequired()
		}
	}
	return formatError("Please select a valid option") + "\n\n" + formatAddPaymentMethodEmailRequired()
}

func (s *Service) handleBillingAdd(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	selectedNumber, hasNum := parseIntTrim(messageBody)
	if hasNum {
		switch selectedNumber {
		case 1:
			// TODO: PayStack payment link generation. Simulate success for now.
			newState := state
			newState.Step = "billing_added"
			newState.PreviousStep = "billing_add"
			s.updateConversationState(ctx, chatID, newState)
			return formatPaymentMethodAdded()
		case 2:
			paymentMethods := s.getCustomerPaymentMethods(ctx, customerID)
			newState := state
			newState.Step = "billing_list"
			newState.PreviousStep = "billing_add"
			s.updateConversationState(ctx, chatID, newState)
			return formatBillingManagement(paymentMethods)
		case 3:
			newState := state
			newState.Step = "main_menu"
			newState.PreviousStep = "billing_add"
			s.updateConversationState(ctx, chatID, newState)
			return formatMainMenu("", 0, 0, "")
		default:
			return formatError("Please select a valid option (1-3)") + "\n\n" + formatAddPaymentMethod()
		}
	}
	return formatError("Please select a valid option") + "\n\n" + formatAddPaymentMethod()
}

func (s *Service) handleBillingAdded(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	_ = messageBody
	paymentMethods := s.getCustomerPaymentMethods(ctx, customerID)
	newState := state
	newState.Step = "billing_list"
	newState.PreviousStep = "billing_added"
	s.updateConversationState(ctx, chatID, newState)
	return formatBillingManagement(paymentMethods)
}
